core/src/ofac_stage.rs

```diff
@@ -0,0 +1,542 @@
+use std::{
+    collections::HashSet,
+    thread,
+    thread::{Builder, JoinHandle},
+};
+
+use crossbeam_channel::{Receiver, Sender};
+use dashmap::DashMap;
+use log::warn;
+use solana_core::banking_stage::BankingPacketBatch;
+use solana_perf::packet::PacketBatch;
+use solana_sdk::{
+    address_lookup_table_account::AddressLookupTableAccount, pubkey::Pubkey,
+    transaction::VersionedTransaction,
+};
+
+pub struct OfacStage {
+    ofac_thread: JoinHandle<()>,
+}
+
+impl OfacStage {
+    pub fn new(
+        verified_receiver: Receiver<BankingPacketBatch>,
+        ofac_sender: Sender<BankingPacketBatch>,
+        ofac_addresses: &HashSet<Pubkey>,
+        address_lookup_table_cache: &DashMap<Pubkey, AddressLookupTableAccount>,
+    ) -> OfacStage {
+        let ofac_thread = if ofac_addresses.is_empty() {
+            Self::spawn_passthrough(verified_receiver, ofac_sender)
+        } else {
+            Self::spawn_ofac_thread(
+                verified_receiver,
+                ofac_sender,
+                ofac_addresses,
+                address_lookup_table_cache,
+            )
+        };
+        OfacStage { ofac_thread }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        self.ofac_thread.join()
+    }
+
+    fn spawn_passthrough(
+        verified_receiver: Receiver<BankingPacketBatch>,
+        ofac_sender: Sender<BankingPacketBatch>,
+    ) -> JoinHandle<()> {
+        Builder::new()
+            .name("ofac_stage".into())
+            .spawn(move || {
+                while let Ok(packets) = verified_receiver.recv() {
+                    if ofac_sender.send(packets).is_err() {
+                        warn!("ofac_sender send error, returning early");
+                        return;
+                    }
+                }
+                warn!("verified_receiver receive error, returning early");
+            })
+            .unwrap()
+    }
+
+    fn spawn_ofac_thread(
+        verified_receiver: Receiver<BankingPacketBatch>,
+        ofac_sender: Sender<BankingPacketBatch>,
+        ofac_addresses: &HashSet<Pubkey>,
+        address_lookup_table_cache: &DashMap<Pubkey, AddressLookupTableAccount>,
+    ) -> JoinHandle<()> {
+        let address_lookup_table_cache = address_lookup_table_cache.clone();
```

Comment:
not sure if we want to clone, looks like an independent copy. maybe use Arc?

https://docs.rs/dashmap/latest/src/dashmap/lib.rs.html#93-107